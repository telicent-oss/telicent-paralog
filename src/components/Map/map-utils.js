import * as turf from "@turf/turf";
import { isEmpty } from "lodash";

const filterEmptyElements = (element) => !isEmpty(element);

export const generatePointAssetFeatures = (
  assets,
  dependencies,
  selectedElements,
) => {
  const pointAssets = assets
    .filter((asset) => asset.lat && asset.lng)
    .map((asset) => {
      return asset.createPointAsset();
    })
    .filter(filterEmptyElements);

  const pointAssetDependencies = dependencies
    .map((dependency) => {
      return dependency.createLineFeature(assets, selectedElements);
    })
    .filter(filterEmptyElements);
  return [...pointAssetDependencies, ...pointAssets];
};

export const generateLinearAssetFeatures = (assets, selectedElements) => {
  const linearAssets = assets
    .filter((asset) => !isEmpty(asset.geometry))
    .map((asset) => asset.createLinearAsset(selectedElements))
    .filter(filterEmptyElements);

  return linearAssets;
};

export const fitMultiToBounds = (map, selectedElements, assets) => {
  if (isEmpty(selectedElements) || isEmpty(assets)) return;

  const lngLats = selectedElements
    .filter((element) => element?.isPointAsset && element?.hasLatLng)
    .map((element) => [element.lng, element.lat]);
  const pointAssetMPFeature = turf.multiPoint(lngLats);

  const linearAssetGeometry = selectedElements
    .filter((element) => element?.isLinearAsset)
    .map((element) => element.createSegmentCoords());
  const linearAssetMLSFeature = turf.multiLineString(linearAssetGeometry);

  const dependencyGeometry = selectedElements
    .filter((element) => element?.isDependency)
    .map((element) => element.generateCoordinates(assets));
  const dependencyMLSFeature = turf.multiLineString(dependencyGeometry);

  if (
    isEmpty(pointAssetMPFeature.geometry.coordinates) &&
    isEmpty(linearAssetMLSFeature.geometry.coordinates) &&
    isEmpty(dependencyMLSFeature.geometry.coordinates)
  )
    return;

  const collection = turf.featureCollection([
    pointAssetMPFeature,
    linearAssetMLSFeature,
    dependencyMLSFeature,
  ]);
  const bbox = turf.bbox(collection);

  map.fitBounds(bbox, {
    padding: { top: 15, bottom: 25, left: 15, right: 55 },
  });
};

const fitLineStringToBounds = (map, geometry) => {
  if (isEmpty(geometry)) return;

  const lineString = turf.lineString(geometry);
  const bbox = turf.bbox(lineString);

  map.fitBounds(bbox, {
    padding: { top: 10, bottom: 25, left: 15, right: 5 },
  });
};

export const fitToBounds = (map, selectedElement, assets) => {
  if (selectedElement?.isPointAsset && selectedElement?.hasLatLng) {
    map.flyTo({ center: [selectedElement.lng, selectedElement.lat], zoom: 12 });
    return;
  }

  if (selectedElement?.isLinearAsset) {
    const geometry = selectedElement.createSegmentCoords();
    fitLineStringToBounds(map, geometry);
    return;
  }

  if (selectedElement?.isDependency && !isEmpty(assets)) {
    const geometry = selectedElement.generateCoordinates(assets);
    fitLineStringToBounds(map, geometry);
    return;
  }
};
