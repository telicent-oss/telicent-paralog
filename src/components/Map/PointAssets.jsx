import React, { useEffect, useMemo, useState } from "react";
import { Layer, Marker, Source } from "react-map-gl";
import PropTypes from "prop-types";
import { useOntologyStyles } from "@telicent-oss/ds";

import { findElement } from "utils";
import { FEATURE_COLLECTION, GEOJSON } from "./TelicentMap";
import { generatePointAssetFeatures } from "./map-utils";
import { FLOOD_AREA_LAYERS, heatmap, pointAssetCxnLayer } from "./layers";

const PointAssets = ({ map, assets, dependencies, selectedElements, onElementClick, moveTo }) => {
  const stableAssetsRef = useMemo(() => assets ?? [], [assets]);
  const stableDependenciesRef = useMemo(() => dependencies ?? [], [dependencies]);
  const stableElementsRef = useMemo(() => selectedElements ?? [], [selectedElements]);

  const [features, setFeatures] = useState([]);
  const points = features.filter((feature) => feature.geometry.type === "Point");

  useEffect(() => {
    const features = generatePointAssetFeatures(stableAssetsRef, stableDependenciesRef, stableElementsRef);
    setFeatures(features);
  }, [stableAssetsRef, stableDependenciesRef, stableElementsRef]);

  useEffect(() => {
    if (!map) return;

    const onLoad = () => {
      map.getMap().moveLayer(heatmap.id, FLOOD_AREA_LAYERS[0].id);
    };

    map.on("load", onLoad);

    return () => {
      map.off("load", onLoad);
    };
  }, [map]);

  const handleOnAssetClick = (event, clickedFeature) => {
    const { originalEvent } = event;
    const isMultiSelect = originalEvent.shiftKey;
    originalEvent.stopPropagation();

    const connectedDependencies = features
      .filter((feature) => feature.geometry.type === "LineString")
      .filter((feature) => {
        const isDependent = feature.properties.dependent === clickedFeature.properties.uri;
        const isProvider = feature.properties.provider === clickedFeature.properties.uri;
        return isDependent || isProvider;
      });

    const connectedAssets = connectedDependencies.map((feature) => {
      const isDependent = feature.properties.dependent === clickedFeature.properties.uri;

      if (isDependent) {
        const providerAsset = points.find(
          (point) => point.properties.uri === feature.properties.provider
        );
        return providerAsset;
      }

      const dependentAsset = points.find(
        (point) => point.properties.uri === feature.properties.dependent
      );
      return dependentAsset;
    });

    const elements = [clickedFeature, ...connectedAssets, ...connectedDependencies].map((feature) =>
      findElement([...assets, ...dependencies], feature.properties.uri)
    );

    onElementClick(isMultiSelect, elements);
    moveTo({ cachedElements: stableElementsRef, selectedElements: elements });
  };

  const isSelected = (feature) => {
    const isSelected = stableElementsRef.some(
      (selectedElement) => selectedElement.uri === feature.properties.uri
    );
    return isSelected;
  };

  return (
    <Source
      id="point-assets"
      type={GEOJSON}
      data={{ type: FEATURE_COLLECTION, features }}
      generateId
    >
      <Layer {...heatmap} />
      <Layer {...pointAssetCxnLayer} />
      <AssetIcons features={points} onAssetClick={handleOnAssetClick} isSelected={isSelected} />
    </Source>
  );
};

PointAssets.propTypes = {
  assets: PropTypes.array,
  dependencies: PropTypes.array,
  selectedElements: PropTypes.array,
  onElementClick: PropTypes.func,
  moveTo: PropTypes.func,
};

export default PointAssets;

const AssetIcons = ({ features, isSelected, onAssetClick }) => {
  const { findIcon } = useOntologyStyles();

  return features.map((feature) => {
    const { geometry, properties } = feature;
    const iconStyles = findIcon(properties.type);
    return (
      <Marker
        key={properties.uri}
        id={properties.uri}
        longitude={geometry.coordinates[0]}
        latitude={geometry.coordinates[1]}
        onClick={(event) => onAssetClick(event, feature)}
        style={{
          backgroundColor: iconStyles.backgroundColor,
          color: iconStyles.color,
          borderColor: isSelected(feature) ? "#fff" : iconStyles.color,
          borderRadius: "9999px",
          cursor: "pointer",
          borderWidth: "2px",
        }}
      >
        {/* Need to add this back in once the icon data is fixed */}
        {/* {iconStyles?.faIcon ? (
          <i className={classNames(iconStyles.faIcon, "px-1")} />
        ) : ( */}
        <p className="p-1 font-body">{iconStyles.iconFallbackText}</p>
        {/* )} */}
      </Marker>
    );
  });
};
