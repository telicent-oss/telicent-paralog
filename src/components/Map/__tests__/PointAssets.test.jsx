import { screen } from "@testing-library/react";
import {
  LARGE_WIND_FARMS_AND_COAL_PLANT_COMPLEX_DEPENDENCIES,
  LARGE_WIND_FARM_ASSETS,
  COAL_PLANT_COMPLEX_ASSETS,
} from "mocks";
import {
  DSProvidersWrapper,
  getCreatedAssets,
  getCreatedDependencies,
  renderWithQueryClient,
} from "test-utils";
import * as mapUtils from "../map-utils";
import PointAssets from "../PointAssets";

describe("Point asset component", () => {

  const getLineStringFeatures = (feature) => feature.geometry.type === "LineString";
  const getPointFeatures = (feature) => feature.geometry.type === "Point";

  const renderPointAssets = (pointAssetProps) => {
    return renderWithQueryClient(<PointAssets {...pointAssetProps} />, {
      wrapper: DSProvidersWrapper,
    });
  };
  test("should generate empty features when there are no elements", () => {
    const spyOnGeneratePointAssetFeatures = jest.spyOn(mapUtils, "generatePointAssetFeatures");
    renderPointAssets();
    expect(spyOnGeneratePointAssetFeatures).toHaveReturnedWith([]);
  });

  test("should generate asset features", async () => {
    const spyOnGeneratePointAssetFeatures = jest.spyOn(mapUtils, "generatePointAssetFeatures");
    const createdAssets = await getCreatedAssets(
      [
        ...LARGE_WIND_FARM_ASSETS,
        ...COAL_PLANT_COMPLEX_ASSETS,
      ],
      ["E001", "E003"]
    );

    renderPointAssets({ assets: createdAssets });

    expect(spyOnGeneratePointAssetFeatures).toBeCalledTimes(1);
    expect(spyOnGeneratePointAssetFeatures.mock.results[0].value).toHaveLength(2);
    expect(spyOnGeneratePointAssetFeatures.mock.results[0].value).toMatchSnapshot();
    expect(screen.getByTestId("https://www.example.com/Instruments#E003")).toBeInTheDocument();
    expect(screen.getByTestId("https://www.example.com/Instruments#E001")).toBeInTheDocument();
  });

  test("should generate dependency features", async () => {
    const spyOnGeneratePointAssetFeatures = jest.spyOn(mapUtils, "generatePointAssetFeatures");
    const createdAssets = await getCreatedAssets(
      [
        ...LARGE_WIND_FARM_ASSETS,
        ...COAL_PLANT_COMPLEX_ASSETS,
      ],
      ["E001", "E003"]
    );

    const createdDependencies = getCreatedDependencies(
      LARGE_WIND_FARMS_AND_COAL_PLANT_COMPLEX_DEPENDENCIES,
      ["E001 - E003"]
    );
    renderPointAssets({ assets: createdAssets, dependencies: createdDependencies });

    expect(spyOnGeneratePointAssetFeatures).toBeCalledTimes(1);
    expect(spyOnGeneratePointAssetFeatures.mock.results[0].value).toHaveLength(3);
    expect(spyOnGeneratePointAssetFeatures.mock.results[0].value).toMatchSnapshot();
  });

  test("should NOT generate dependencies when there are no assets", () => {
    const spyOnGeneratePointAssetFeatures = jest.spyOn(mapUtils, "generatePointAssetFeatures");
    const createdDependencies = getCreatedDependencies(
      LARGE_WIND_FARMS_AND_COAL_PLANT_COMPLEX_DEPENDENCIES,
      ["E001 - E003"]
    );
    renderPointAssets({ dependencies: createdDependencies });

    expect(spyOnGeneratePointAssetFeatures).toBeCalledTimes(1);
    expect(spyOnGeneratePointAssetFeatures).toHaveReturnedWith([]);
  });

  test("should NOT generate features for selected elements which don't exist", async () => {
    const spyOnGeneratePointAssetFeatures = jest.spyOn(mapUtils, "generatePointAssetFeatures");
    const assets = await getCreatedAssets(LARGE_WIND_FARM_ASSETS, [
      "E003",
      "E025",
    ]);
    const selectedElements = await getCreatedAssets(COAL_PLANT_COMPLEX_ASSETS, [
      "E001",
    ]);
    renderPointAssets({ assets, selectedElements });

    expect(spyOnGeneratePointAssetFeatures).toBeCalledTimes(1);
    expect(spyOnGeneratePointAssetFeatures.mock.results[0].value).toHaveLength(2);
    expect(spyOnGeneratePointAssetFeatures.mock.results[0].value).toMatchSnapshot();
  });

  test("should NOT generate point asset features when previously selected assets are removed", async () => {
    const spyOnGeneratePointAssetFeatures = jest.spyOn(mapUtils, "generatePointAssetFeatures");
    const assets = await getCreatedAssets(LARGE_WIND_FARM_ASSETS, [
      "E003",
      "E025",
    ]);

    const { rerender } = renderPointAssets({ assets, selectedElements: assets });

    expect(spyOnGeneratePointAssetFeatures).toBeCalledTimes(1);
    expect(spyOnGeneratePointAssetFeatures.mock.results[0].value).toHaveLength(2);

    rerender(<PointAssets assets={[]} selectedElements={assets} />, {
      wrapper: DSProvidersWrapper,
    });
    expect(spyOnGeneratePointAssetFeatures).toBeCalledTimes(2);
    expect(spyOnGeneratePointAssetFeatures).toHaveReturnedWith([]);
  });

  test("should generate selected point asset features for selected assets which exist", async () => {
    const spyOnGeneratePointAssetFeatures = jest.spyOn(mapUtils, "generatePointAssetFeatures");
    const assets = await getCreatedAssets(
      [
        ...LARGE_WIND_FARM_ASSETS,
        ...COAL_PLANT_COMPLEX_ASSETS,
      ],
      ["E001", "E025", "E003"]
    );
    const { rerender } = renderPointAssets({ assets, selectedElements: assets });

    expect(spyOnGeneratePointAssetFeatures).toBeCalledTimes(1);
    expect(spyOnGeneratePointAssetFeatures.mock.results[0].value).toHaveLength(3);

    const filteredAssets = assets.filter((asset) => {
      const typeFilter = "http://ies.example.com/ontology/ies#CoalPlantComplex";
      const isCoalPlantAsset = asset.type === typeFilter;
      return isCoalPlantAsset;
    });
    rerender(<PointAssets assets={filteredAssets} selectedElements={assets} />, {
      wrapper: DSProvidersWrapper,
    });

    expect(spyOnGeneratePointAssetFeatures).toBeCalledTimes(2);
    expect(spyOnGeneratePointAssetFeatures.mock.results[1].value).toHaveLength(1);
  });

  test("should NOT generate dependency features when previously dependencies when assets are removed", async () => {
    const spyOnGeneratePointAssetFeatures = jest.spyOn(mapUtils, "generatePointAssetFeatures");
    const assets = await getCreatedAssets(
      [
        ...LARGE_WIND_FARM_ASSETS,
        ...COAL_PLANT_COMPLEX_ASSETS,
      ],
      ["E001", "E003"]
    );
    const dependencies = getCreatedDependencies(
      LARGE_WIND_FARMS_AND_COAL_PLANT_COMPLEX_DEPENDENCIES,
      ["E001 - E003"]
    );
    const { rerender } = renderPointAssets({
      assets,
      dependencies,
      selectedElements: dependencies,
    });

    expect(spyOnGeneratePointAssetFeatures).toBeCalledTimes(1);
    const results =
      spyOnGeneratePointAssetFeatures.mock.results[0].value.filter(getLineStringFeatures);
    expect(results).toHaveLength(1);
    expect(results[0].properties.selected).toBe(true);

    rerender(<PointAssets assets={[]} dependencies={[]} selectedElements={dependencies} />, {
      wrapper: DSProvidersWrapper,
    });
    expect(spyOnGeneratePointAssetFeatures).toBeCalledTimes(2);
    expect(spyOnGeneratePointAssetFeatures).toHaveReturnedWith([]);
  });

  test("should generate selected dependency features for assets which exist", async () => {
    const spyOnGeneratePointAssetFeatures = jest.spyOn(mapUtils, "generatePointAssetFeatures");
    const assets = await getCreatedAssets(
      [
        ...LARGE_WIND_FARM_ASSETS,
        ...COAL_PLANT_COMPLEX_ASSETS,
      ],
      ["E001", "E025", "E003"]
    );
    const dependencies = getCreatedDependencies(
      LARGE_WIND_FARMS_AND_COAL_PLANT_COMPLEX_DEPENDENCIES,
      ["E001 - E003", "E003 - E025"]
    );
    const { rerender } = renderPointAssets({
      assets,
      dependencies,
      selectedElements: dependencies,
    });

    expect(spyOnGeneratePointAssetFeatures).toBeCalledTimes(1);
    let results =
      spyOnGeneratePointAssetFeatures.mock.results[0].value.filter(getLineStringFeatures);
    expect(results).toHaveLength(2);
    expect(results[0].properties.selected).toBe(true);
    expect(results[1].properties.selected).toBe(true);

    const filteredAssets = assets.filter((asset) => {
      const typeFilter =
        "http://ies.example.com/ontology/ies#LargeWindFarm";
      const isLargeWindFarmAsset = asset.type === typeFilter;
      return isLargeWindFarmAsset;
    });

    rerender(
      <PointAssets
        assets={filteredAssets}
        dependencies={dependencies}
        selectedElements={dependencies}
      />,
      { wrapper: DSProvidersWrapper }
    );
    expect(spyOnGeneratePointAssetFeatures).toBeCalledTimes(2);
    results = spyOnGeneratePointAssetFeatures.mock.results[1].value.filter(getLineStringFeatures);
    expect(results).toHaveLength(1);
    expect(results[0].properties.selected).toBe(true);
  });

  test("should NOT generate point asset and dependency features when previoulsy selected elements are removed", async () => {
    const spyOnGeneratePointAssetFeatures = jest.spyOn(mapUtils, "generatePointAssetFeatures");
    const assets = await getCreatedAssets(
      [
        ...LARGE_WIND_FARM_ASSETS,
        ...COAL_PLANT_COMPLEX_ASSETS,
      ],
      ["E001", "E003"]
    );
    const dependencies = getCreatedDependencies(
      LARGE_WIND_FARMS_AND_COAL_PLANT_COMPLEX_DEPENDENCIES,
      ["E001 - E003"]
    );
    const { rerender } = renderPointAssets({
      assets,
      dependencies,
      selectedElements: [...assets, ...dependencies],
    });

    expect(spyOnGeneratePointAssetFeatures).toBeCalledTimes(1);
    const assetFeatures =
      spyOnGeneratePointAssetFeatures.mock.results[0].value.filter(getPointFeatures);
    const dependenciesFeatures =
      spyOnGeneratePointAssetFeatures.mock.results[0].value.filter(getLineStringFeatures);
    // Point assets
    expect(assetFeatures).toHaveLength(2);

    // Point asset dependencies
    expect(dependenciesFeatures).toHaveLength(1);
    expect(dependenciesFeatures[0].properties.selected).toBe(true);

    rerender(
      <PointAssets assets={[]} dependencies={[]} selectedElements={[...assets, ...dependencies]} />,
      { wrapper: DSProvidersWrapper }
    );
    expect(spyOnGeneratePointAssetFeatures).toBeCalledTimes(2);
    expect(spyOnGeneratePointAssetFeatures).toHaveReturnedWith([]);
  });

  test("should generate selected point asset and dependency features for elements which exist", async () => {
    const spyOnGeneratePointAssetFeatures = jest.spyOn(mapUtils, "generatePointAssetFeatures");
    const assets = await getCreatedAssets(
      [
        ...LARGE_WIND_FARM_ASSETS,
        ...COAL_PLANT_COMPLEX_ASSETS,
      ],
      ["E001", "E025", "E003"]
    );
    const dependencies = getCreatedDependencies(
      LARGE_WIND_FARMS_AND_COAL_PLANT_COMPLEX_DEPENDENCIES,
      ["E001 - E003", "E003 - E025"]
    );
    const { rerender } = renderPointAssets({
      assets,
      dependencies,
      selectedElements: [...assets, ...dependencies],
    });

    let assetFeatures =
      spyOnGeneratePointAssetFeatures.mock.results[0].value.filter(getPointFeatures);
    let dependenciesFeatures =
      spyOnGeneratePointAssetFeatures.mock.results[0].value.filter(getLineStringFeatures);

    expect(spyOnGeneratePointAssetFeatures).toBeCalledTimes(1);

    // Point assets
    expect(assetFeatures).toHaveLength(3);

    // Point asset dependencies
    expect(dependenciesFeatures).toHaveLength(2);
    expect(dependenciesFeatures[0].properties.selected).toBe(true);
    expect(dependenciesFeatures[1].properties.selected).toBe(true);

    const typeFilter = "http://ies.example.com/ontology/ies#CoalPlantComplex";
    const filteredAssets = assets.filter((asset) => {
      const isCoalPlantAsset = asset.type === typeFilter;
      return !isCoalPlantAsset;
    });
    const filteredDependencies = dependencies.filter((dependency) => {
      const isCoalPlantAsset =
        dependency.dependent.type === typeFilter || dependency.provider.type === typeFilter;
      return !isCoalPlantAsset;
    });
    rerender(
      <PointAssets
        assets={filteredAssets}
        dependencies={filteredDependencies}
        selectedElements={[...assets, ...dependencies]}
      />,
      { wrapper: DSProvidersWrapper }
    );

    assetFeatures = spyOnGeneratePointAssetFeatures.mock.results[1].value.filter(getPointFeatures);
    dependenciesFeatures =
      spyOnGeneratePointAssetFeatures.mock.results[1].value.filter(getLineStringFeatures);

    expect(spyOnGeneratePointAssetFeatures).toBeCalledTimes(2);
    // Point assets
    expect(assetFeatures).toHaveLength(2);

    // Point asset dependencies
    expect(dependenciesFeatures).toHaveLength(1);
    expect(dependenciesFeatures[0].properties.selected).toBe(true);
  });
});
