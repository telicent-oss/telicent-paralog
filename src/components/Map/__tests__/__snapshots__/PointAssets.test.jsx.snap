// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Point asset component > should NOT generate features for selected elements which don't exist 1`] = `
[
  {
    "geometry": {
      "coordinates": [
        -1231231,
        1.111111,
      ],
      "type": "Point",
    },
    "properties": {
      "criticality": 2,
      "id": "E003",
      "type": "http://ies.example.com/ontology/ies#LargeWindFarm",
      "uri": "https://www.example.com/Instruments#E003",
    },
    "type": "Feature",
  },
  {
    "geometry": {
      "coordinates": [
        100.098765,
        1.555555,
      ],
      "type": "Point",
    },
    "properties": {
      "criticality": 2,
      "id": "E025",
      "type": "http://ies.example.com/ontology/ies#LargeWindFarm",
      "uri": "https://www.example.com/Instruments#E025",
    },
    "type": "Feature",
  },
]
`;

exports[`Point asset component > should generate asset features 1`] = `
[
  {
    "geometry": {
      "coordinates": [
        -1231231,
        1.111111,
      ],
      "type": "Point",
    },
    "properties": {
      "criticality": 2,
      "id": "E003",
      "type": "http://ies.example.com/ontology/ies#LargeWindFarm",
      "uri": "https://www.example.com/Instruments#E003",
    },
    "type": "Feature",
  },
  {
    "geometry": {
      "coordinates": [
        10.1233333,
        100.123123,
      ],
      "type": "Point",
    },
    "properties": {
      "criticality": 1,
      "id": "E001",
      "type": "http://ies.example.com/ontology/ies#CoalPlantComplex",
      "uri": "https://www.example.com/Instruments#E001",
    },
    "type": "Feature",
  },
]
`;

exports[`Point asset component > should generate dependency features 1`] = `
[
  {
    "geometry": {
      "coordinates": [
        [
          10.1233333,
          100.123123,
        ],
        [
          -1231231,
          1.111111,
        ],
      ],
      "type": "LineString",
    },
    "properties": {
      "criticality": 2,
      "dependent": "https://www.example.com/Instruments#E001",
      "id": "E001 - E003",
      "lineColor": "#ffb60a",
      "lineOpacity": 0.3,
      "provider": "https://www.example.com/Instruments#E003",
      "selected": false,
      "uri": "https://www.example.com/Instruments#_E001_E003_dependency",
    },
    "type": "Feature",
  },
  {
    "geometry": {
      "coordinates": [
        -1231231,
        1.111111,
      ],
      "type": "Point",
    },
    "properties": {
      "criticality": 2,
      "id": "E003",
      "type": "http://ies.example.com/ontology/ies#LargeWindFarm",
      "uri": "https://www.example.com/Instruments#E003",
    },
    "type": "Feature",
  },
  {
    "geometry": {
      "coordinates": [
        10.1233333,
        100.123123,
      ],
      "type": "Point",
    },
    "properties": {
      "criticality": 1,
      "id": "E001",
      "type": "http://ies.example.com/ontology/ies#CoalPlantComplex",
      "uri": "https://www.example.com/Instruments#E001",
    },
    "type": "Feature",
  },
]
`;
